ICSharpCode.Decompiler/TypeSystem/Implementation/LocalFunctionMethod.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Copyright (c) 2019 Siegfried Pammer
+// 
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this
+// software and associated documentation files (the "Software"), to deal in the Software
+// without restriction, including without limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons
+// to whom the Software is furnished to do so, subject to the following conditions:
+// 
+// The above copyright notice and this permission notice shall be included in all copies or
+// substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+// PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+// FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using ICSharpCode.Decompiler.Util;
+
+namespace ICSharpCode.Decompiler.TypeSystem.Implementation
+{
+	/// <summary>
+	/// A local function has zero or more compiler-generated parameters added at the end.
+	/// </summary>
+	class LocalFunctionMethod : IMethod
+	{
+		readonly IMethod baseMethod;
+
+		public LocalFunctionMethod(IMethod baseMethod, int numberOfCompilerGeneratedParameters)
+		{
+			this.baseMethod = baseMethod;
+			this.NumberOfCompilerGeneratedParameters = numberOfCompilerGeneratedParameters;
+		}
+
+
+		public bool Equals(IMember obj, TypeVisitor typeNormalization)
+		{
+			if (!(obj is LocalFunctionMethod other))
+				return false;
+			return baseMethod.Equals(other.baseMethod, typeNormalization);
+		}
+
+		public override bool Equals(object obj)
+		{
+			var other = obj as LocalFunctionMethod;
+			if (other == null)
+				return false;
+			return baseMethod.Equals(other.baseMethod);
+		}
+		
+		public override int GetHashCode()
+		{
+			unchecked {
+				return baseMethod.GetHashCode() + 1;
+			}
+		}
+
+		public override string ToString()
+		{
+			return string.Format("[LocalFunctionMethod: ReducedFrom={0}, NumberOfGeneratedParameters={1}]", ReducedFrom, NumberOfCompilerGeneratedParameters);
+		}
+
+		internal int NumberOfCompilerGeneratedParameters { get; }
+
+		public IMember MemberDefinition {
+			get {
```

Comment:
I think this property can unconditionally return `this` now that the SpecializedMethod is wrapping the LocalFunctionMethod